File: C++17/백준/Gold/27501. RGB트리/RGB트리.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
using ll = long long;

vector<vector<int>> tree;

void dfs(int node, int parent, vector<array<ll, 3>>& dp) {
    for (int nxt : tree[node]) {
        if (nxt == parent) continue;
        dfs(nxt, node, dp);
        dp[node][0] += max(dp[nxt][1], dp[nxt][2]);
        dp[node][1] += max(dp[nxt][0], dp[nxt][2]);
        dp[node][2] += max(dp[nxt][0], dp[nxt][1]);
    }
}

void trace(int node, int parent, vector<array<ll, 3>>& dp, vector<int>& colors) {
    int parent_color = colors[node];

    for (int nxt : tree[node]) {
        if (nxt == parent) continue;
        int best_color = -1;
        int max_val = -1;

        for (int i = 0; i < 3; i++) {
            if (i == parent_color) continue;
            if (dp[nxt][i] > max_val) {
                max_val = dp[nxt][i];
                best_color = i;
            }
        }
        colors[nxt] = best_color;
        trace(nxt, node, dp, colors);
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int N;
    cin >> N;
    tree.resize(N + 1);
    for (int i = 0; i < N - 1; i++) {
        int a, b; cin >> a >> b;
        tree[a].push_back(b);
        tree[b].push_back(a);
    }

    vector<array<ll, 3>> light(N + 1), dp(N + 1);
    for (int i = 1; i <= N; i++) {
        for (int j = 0; j < 3; j++) {
            int c; cin >> c;
            light[i][j] = c;
        }
    }

    dp = light;
    dfs(1, 0, dp);

    vector<int> colors(N + 1);
    int root_color = 0;
    for (int i = 1; i < 3; i++) {
        if (dp[1][i] > dp[1][root_color]) root_color = i;
    }
    colors[1] = root_color;
    trace(1, 0, dp, colors);

    cout << max({dp[1][0], dp[1][1], dp[1][2]}) << '\n';
    char color_map[3] = {'R', 'G', 'B'};
    for (int i = 1; i <= N; i++) {
        cout << color_map[colors[i]];
    }

    return 0;
}

File: PyPy3/백준/Gold/27501. RGB트리/RGB트리.py
import sys
input = sys.stdin.readline

N = int(input())
graph = [[] for _ in range(N + 1)]
for i in range(N - 1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
    
light = [[0, 0, 0]]
for _ in range(N):
    light.append(list(map(int, input().split())))

dp = [row[:] for row in light]
visited = [False] * (N + 1)
visited[1] = True
def dfs(node):
    for child in graph[node]:
        if not visited[child]:
            visited[child] = True
            dfs(child)
            
            dp[node][0] += max(dp[child][1], dp[child][2])
            dp[node][1] += max(dp[child][0], dp[child][2])
            dp[node][2] += max(dp[child][0], dp[child][1])

dfs(1)

colors = [0] * (N + 1)
root_color = 0
for i in range(1, 3):
    if dp[1][i] > dp[1][root_color]:
        root_color = i 
colors[1] = root_color

def trace(node, parent):
    parent_color = colors[node]
    
    for child in graph[node]:
        if child != parent:
            best_color = -1
            max_val = -1
            
            for c in range(3):
                if c != parent_color:
                    if dp[child][c] > max_val:
                        max_val = dp[child][c]
                        best_color = c
            colors[child] = best_color
            trace(child, node)

trace(1, 0)

print(max(dp[1]))
rgb_map = ['R', 'G', 'B']
rgb = [rgb_map[colors[i]] for i in range(1, N + 1)]
print(*rgb, sep = '')